contracts/src/v0.8/dev/vrf/testhelpers/VRFCoordinatorV2PlusUpgradedVersion.sol

```diff
@@ -0,0 +1,715 @@
+  /// @param subId subscription ID
+  event MigrationCompleted(address newCoordinator, uint256 subId);
+
+  /// @notice emitted when migrate() is called and given coordinator is not registered as migratable target
+  error CoordinatorNotRegistered(address coordinatorAddress);
+
+  /// @notice emitted when migrate() is called and given coordinator is registered as migratable target
+  error CoordinatorAlreadyRegistered(address coordinatorAddress);
+
+  /// @dev encapsulates data to be migrated from current coordinator
+  struct V1MigrationData {
+    uint8 fromVersion;
+    uint256 subId;
+    address subOwner;
+    address[] consumers;
+    uint96 linkBalance;
+    uint96 ethBalance;
+  }
+
+  function isTargetRegistered(address target) internal view returns (bool) {
+    for (uint256 i = 0; i < s_migrationTargets.length; i++) {
+      if (s_migrationTargets[i] == target) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  function registerMigratableCoordinator(address target) external onlyOwner {
+    if (isTargetRegistered(target)) {
+      revert CoordinatorAlreadyRegistered(target);
+    }
+    s_migrationTargets.push(target);
+    emit CoordinatorRegistered(target);
+  }
+
+  function migrate(uint256 subId, address newCoordinator) external nonReentrant {
+    if (!isTargetRegistered(newCoordinator)) {
+      revert CoordinatorNotRegistered(newCoordinator);
+    }
+    (uint96 balance, uint96 ethBalance, , address owner, address[] memory consumers) = getSubscription(subId);
+    require(owner == msg.sender, "Not subscription owner");
+    require(!pendingRequestExists(subId), "Pending request exists");
+
+    V1MigrationData memory migrationData = V1MigrationData({
+      fromVersion: migrationVersion(),
+      subId: subId,
+      subOwner: owner,
+      consumers: consumers,
+      linkBalance: balance,
+      ethBalance: ethBalance
+    });
+    bytes memory encodedData = abi.encode(migrationData);
+    deleteSubscription(subId);
+    IVRFCoordinatorV2PlusMigration(newCoordinator).onMigration{value: ethBalance}(encodedData);
+
+    // Only transfer LINK if the token is active and there is a balance.
+    if (address(LINK) != address(0) && balance != 0) {
+      require(LINK.transfer(address(newCoordinator), balance), "insufficient funds");
+    }
+
+    // despite the fact that we follow best practices this is still probably safest
+    // to prevent any re-entrancy possibilities.
+    s_config.reentrancyLock = true;
+    for (uint256 i = 0; i < consumers.length; i++) {
+      IVRFMigratableConsumerV2Plus(consumers[i]).setCoordinator(newCoordinator);
+    }
+    s_config.reentrancyLock = false;
+
+    emit MigrationCompleted(newCoordinator, subId);
+  }
+
+  function migrationVersion() public pure returns (uint8 version) {
+    return 1;
+  }
+
+  /**
+   * @inheritdoc IVRFCoordinatorV2PlusMigration
+   */
+  function onMigration(bytes calldata encodedData) external payable override {
```

Comment:
I think we should also increment `s_totalBalance` and `s_totalEthBalance`

Reply:
now I had a second thought on this :
should we also update following:
* `s_withdrawableTokens`, `s_withdrawableEth` with keyhash and corresponding oracle
* update `s_subscriptions` with sub id

Reply:
After another thought - we dont need to update `s_withdrawableTokens`, `s_withdrawableEth` as these are updated at fulfilment phase

Reply:
hmm, not sure why, but compilation fails if adding `s_subIds.add(migrationData.subId)`:

```
Error: Member "add" not found or not visible after argument-dependent lookup in struct EnumerableSet.UintSet storage ref
```

`s_subIds` has internal visibility so it should allow to do it, right?